src/statsmodels/formula.jl

```diff
@@ -485,7 +485,16 @@ creating the model matrix.
         error("Could not construct model matrix. Resulting matrix has 0 columns.")
     end
 
-    ModelMatrix{T}(reduce(hcat, blocks), assign)
+    I = size(dfrm, 1)
+    J = mapreduce(x -> size(x, 2), +, blocks)
+    X = Array(eltype(T), I, J)
```

Comment:
Following up on the question in #1045, it would be good to avoid preallocating a dense matrix that's potentially large and expected to be sparse. Uninitialized sparse arrays are tricky though... `spzeros` is probably the best option, but that doesn't generalize well for the dense case. The only thing I can think of off the top of my head would be something not-so-elegant like

```julia
T <: AbstractSparseArray ? spzeros(Tv, Ti, I, J) : Array(Tv, I, J)
```